### PR TITLE
[Speedy] Inline LedgerMode in the Machine

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
We can summarize this big change as follows:
```

LedgerMode=OnLedger|OffLedger
LedgerMode/Machine
-->
OnLedgerMachine|OffLedgerMachine
```

And the motivation is this:

A `Machine` is just a bucket of mutable data. The details of this mutable data determine what basic ops can be executed on the machine. We have two distinct types of machine (on/off ledger).

The basic execution model of a machine, as defined by the syntactic expressions and the runtime continuations, is invariant across both/all machine types.

We want to model this complexity better at the scala type level.

### CODE REVIEW

**Review by nickchapman-da — COMMENTED** [review]
Wow. This is a big change!

My understanding is that the core of this change is:

```
LedgerMode=OnLedger|OffLedger
LedgerMode/Machine
-->
OnLedgerMachine|OffLedgerMachine
```

And the motivation is this:

A `Machine` is just a bucket of mutable data. The details of this mutable data determine what basic ops can be executed on the machine.
We have two distinct types of machine (on/off ledger).

The basic execution model of a machine, as defined by the syntactic expressions and the runtime continuations, is invariant across both/all machine types.

We want to model this complexity better at the scala type level.

**Comment by remyhaemmerle-da** [comment]
> Wow. This is a big change!

It is, (I really try to make the changes as small as possible). 
However, I am quit happy to have reduce the total number of lines. 

You summarize the motivation very well.

**Comment by nickchapman-da** [comment]
> > Wow. This is a big change!
> 
> It is, (I really try to make the changes as small as possible). However, I am quit happy to have reduce the total number of lines.
> 
> You summarize the motivation very well.

Feel free to copy it to the PR description :)

**Review by nickchapman-da — APPROVED** [review]

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 1]
```diff
@@ -1511,18 +1511,17 @@ private[lf] object Speedy {
...
       machine.restoreBase(savedBase)
       machine.restoreFrameAndActuals(frame, actuals)
     }
 
-    override def execute(v: SValue): Control = {
-      restore()
-      machine.withOnLedger("KTryCatchHandler") { onLedger =>
-        onLedger.ptx = onLedger.ptx.endTry
+    override def execute(v: SValue): Control =
+      machine.asOnLedger("KTryCatchHandler") { machine =>
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:1520]: Based on elsewhere, should `"KTryCatchHandler"` be replaced by `productPrefix`?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala [hunk 2]
```diff
@@ -34,16 +34,16 @@ import com.daml.scalautil.Statement.discard
   */
 
 @SuppressWarnings(Array("org.wartremover.warts.Any"))
-object SExpr {
+private[lf] object SExpr {
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala:37]: :+1:

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 3]
```diff
@@ -1555,52 +1554,53 @@ private[lf] object Speedy {
     * producing a text message.
     */
   private[speedy] def unwindToHandler(machine: Machine, excep: SValue.SAny): Control = {
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:1556]: :+1:

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/ScenarioRunner.scala [hunk 4]
```diff
@@ -125,7 +124,7 @@ final class ScenarioRunner private (
   }
 }
 
-object ScenarioRunner {
+private[lf] object ScenarioRunner {
 
   def run(
       buildMachine: () => Speedy.Machine,
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/ScenarioRunner.scala:130]: Good catch.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala]: ff155b94f8661cc0f28d255ae0cc0d8f525fb018

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala]: ff155b94f8661cc0f28d255ae0cc0d8f525fb018

[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala]: ff155b94f8661cc0f28d255ae0cc0d8f525fb018

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/ScenarioRunner.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/ScenarioRunner.scala]: ff155b94f8661cc0f28d255ae0cc0d8f525fb018

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 5]
```diff
@@ -928,51 +1044,6 @@ private[lf] object Speedy {
...
-        traceLog = traceLog,
-        warningLog = warningLog,
-        loggingContext = loggingContext,
-        compiledPackages = compiledPackages,
-      )
-    }
-
     @throws[PackageNotFound]
     @throws[CompilationError]
     // Construct a machine for running an update expression (testing -- avoiding scenarios)
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:1049]: Return type to `fromUpdateExpr` should be `OnLedgerMachine`

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 6]
```diff
@@ -1053,14 +1124,13 @@ private[lf] object Speedy {
...
+      new OffLedgerMachine(
         sexpr = expr,
-        ledgerMode = OffLedger,
         traceLog = traceLog,
         warningLog = warningLog,
-        loggingContext = loggingContext,
         compiledPackages = compiledPackages,
+        profile = new Profile(),
       )
     }
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:1136]: Return type of `fromPureExpr` should be `OffLedgerMachine`

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 7]
```diff
@@ -128,21 +129,70 @@ private[lf] object Speedy {
...
-      limits: interpretation.Limits,
-      disclosureKeyTable: DisclosedContractKeyTable,
-  ) extends LedgerMode {
+      val commitLocation: Option[Location],
+      val limits: interpretation.Limits,
+      val disclosureKeyTable: DisclosedContractKeyTable,
+  )(implicit loggingContext: LoggingContext)
+      extends Machine {
+
+    /** unwindToHandler is called when an exception is thrown by the builtin SBThrow or
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:154]: IIRC, Scaladoc does allow a `/** @inheritdoc */` style document comment to be used here. Suggesting this possibility as you may want to ensure the inherited comment and source comment remain in sync?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 8]
```diff
@@ -1169,7 +1164,7 @@ private[lf] object SBuiltin {
           Control.Question(
             SResultNeedContract(
               coid,
-              onLedger.committers,
+              machine.committers,
               callback = { res =>
                 val control = continue(res)
                 machine.setControl(control)
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:1170]: At lines 1213 and 1231, we can rewrite `record @ _` to simply `_`

**Review by carlpulley-da — APPROVED** [review]
Other than a couple of small niggles, LGTM.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 7 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:154]: The comment make sens only for the onLedger Machine. I remove from the declaration in `Machine`.

**Pull request merged**